node/cmd/ccq/utils.go

```diff
@@ -0,0 +1,231 @@
+package ccq
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/certusone/wormhole/node/pkg/common"
+	gossipv1 "github.com/certusone/wormhole/node/pkg/proto/gossip/v1"
+	"github.com/certusone/wormhole/node/pkg/query"
+	"github.com/wormhole-foundation/wormhole/sdk/vaa"
+	"go.uber.org/zap"
+
+	ethAbi "github.com/certusone/wormhole/node/pkg/watchers/evm/connectors/ethabi"
+	ethBind "github.com/ethereum/go-ethereum/accounts/abi/bind"
+	eth_common "github.com/ethereum/go-ethereum/common"
+	ethCrypto "github.com/ethereum/go-ethereum/crypto"
+	ethClient "github.com/ethereum/go-ethereum/ethclient"
+	ethRpc "github.com/ethereum/go-ethereum/rpc"
+)
+
+func FetchCurrentGuardianSet(rpcUrl, coreAddr string) (*common.GuardianSet, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+	ethContract := eth_common.HexToAddress(coreAddr)
+	rawClient, err := ethRpc.DialContext(ctx, rpcUrl)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to ethereum")
+	}
+	client := ethClient.NewClient(rawClient)
+	caller, err := ethAbi.NewAbiCaller(ethContract, client)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create caller")
+	}
+	currentIndex, err := caller.GetCurrentGuardianSetIndex(&ethBind.CallOpts{Context: ctx})
+	if err != nil {
+		return nil, fmt.Errorf("error requesting current guardian set index: %w", err)
+	}
+	gs, err := caller.GetGuardianSet(&ethBind.CallOpts{Context: ctx}, currentIndex)
+	if err != nil {
+		return nil, fmt.Errorf("error requesting current guardian set value: %w", err)
+	}
+	return &common.GuardianSet{
+		Keys:  gs.Keys,
+		Index: currentIndex,
+	}, nil
+}
+
+type Config struct {
+	Permissions []User `json:"Permissions"`
+}
+
+type User struct {
+	UserName      string        `json:"userName"`
+	ApiKey        string        `json:"apiKey"`
+	AllowUnsigned bool          `json:"allowUnsigned"`
+	AllowedCalls  []AllowedCall `json:"allowedCalls"`
+}
+
+type AllowedCall struct {
+	EthCall *EthCall `json:"ethCall"`
+}
+
+type EthCall struct {
+	Chain           int    `json:"chain"`
+	ContractAddress string `json:"contractAddress"`
+	Call            string `json:"call"`
+}
+
+type Permissions map[string]*permissionEntry
+
+type permissionEntry struct {
+	userName      string
+	apiKey        string
+	allowUnsigned bool
+	allowedCalls  allowedCallsForUser // Key is something like "ethCall:2:000000000000000000000000b4fbf271143f4fbf7b91a5ded31805e42b2208d6:06fdde03"
+}
+
+type allowedCallsForUser map[string]struct{}
+
+// parseConfig parses the permissions config file into a map keyed by API key.
+func parseConfig(fileName string) (Permissions, error) {
```

Comment:
I realize it is a touch more work on the caller, but if you would be so kind as to make this instead `parseConfig(reader io.Reader) (Permissions, error) {`, it would be [much easier to unit test](https://endler.dev/2018/go-io-testing/).

If you want help writing some of these tests I'd be willing to help you do it.

Reply:
It was simple to create two separate methods `parseConfig(buf []byte)` and `parseConfigFile(fileName string)` where the second just calls the first. That made it simple to add parsing unit tests, which I did.

Reply:
Legend. Thanks so much.